src/actions/write/tick.ts

```diff
@@ -211,6 +224,51 @@ export function tickGatewayRegistry({
   };
 }
 
+export function tickVaults({
+  currentBlockHeight,
+  vaults,
+  balances,
+}: {
+  currentBlockHeight: BlockHeight;
+  vaults: DeepReadonly<Vaults>;
+  balances: DeepReadonly<Balances>;
+}): Pick<IOState, 'vaults' | 'balances'> {
+  const updatedBalances: { [address: string]: number } = {};
+  const updatedVaults = Object.keys(vaults).reduce(
+    (acc: Vaults, address: WalletAddress) => {
+      const activeVaults = vaults[address].filter((vault: TokenVault) => {
+        if (vault.end <= currentBlockHeight.valueOf()) {
+          // Initialize the balance if it hasn't been yet
+          if (!updatedBalances[address]) {
+            updatedBalances[address] = balances[address] ?? 0;
+          }
+          // Unlock the vault and update the balance
+          incrementBalance(updatedBalances, address, vault.balance);
+          return false;
+        }
+        return true;
+      });
+
+      if (activeVaults.length > 0) {
+        // Only add to the accumulator if there are active vaults remaining
+        acc[address] = activeVaults;
+      }
+      return acc;
+    },
+    {},
+  );
+
+  // avoids copying balances if not necessary
+  const newBalances: Balances = Object.keys(updatedBalances).length
+    ? { ...balances, ...updatedBalances }
+    : balances;
```

Comment:
@matibat - the intent here is we don't want to copy the `balances` object if we do not need to - if no updates are necessary we return the existing referenced object instead of creating a new one